runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -54,49 +58,119 @@ void waitForWorkflow(UUID workflow, ManagementClient client, Duration timeout,
-            String orchestratorEndpoint = logServers.get(logServers.size() - 1);
-            ManagementClient client = runtime.getRouter(orchestratorEndpoint)
-                    .getClient(ManagementClient.class);
-
+            Request req;
             try {
-                CreateWorkflowResponse resp = client.addNodeRequest(endpointToAdd);
-                log.info("addNode: Workflow id {} for {}", resp.workflowId, endpointToAdd);
-                waitForWorkflow(resp.getWorkflowId(), client, timeout, pollPeriod);
-                Layout newLayout = null;
+                runtime.invalidateLayout();
+                req = requestFunction.apply(runtime.getLayoutView().getLayout());
+
+                waitForWorkflow(req.getId(), req.getClient(), timeout, pollPeriod);
+
                 for (int y = 0; y < runtime.getParameters().getInvalidateRetry(); y++) {
                     runtime.invalidateLayout();
-                    newLayout = runtime.getLayoutView().getLayout();
-                    if (newLayout.getAllServers().contains(endpointToAdd) &&
-                            newLayout.getSegmentsForEndpoint(endpointToAdd).size() == 1) {
-                        log.info("addNode: Successfully added {}", endpointToAdd);
+                    if (validationFunction.apply(runtime
+                            .getLayoutView().getLayout())) {
+                        // Validation has succeeded
                         return;
                     }
+                    log.warn("runWorkflow: failed verification for {} on try {}", req.getId(), y);
                 }
-                log.warn("addNode: Couldn't find {} in {}, retrying...", endpointToAdd, newLayout, x);
-
-            } catch (NetworkException | TimeoutException  e) {
-                log.warn("addNode: Error while requesting to add {}", endpointToAdd, e);
+            } catch (NetworkException | TimeoutException e) {
+                log.warn("runWorkflow: Error ", e);
                 continue;
             }
+            log.warn("runWorkflow: Failed to run {} on try {}", req == null ? null : req, x);
         }
 
-        log.warn("addNode: Failed to add node {}", endpointToAdd);
-        throw new WorkflowException("Failed to add " + endpointToAdd);
+        throw new WorkflowResultUnknownException();
+    }
+
+    /**
+     * Remove a node from the cluster.
+     *
+     * @param endpointToRemove Endpoint of the node to be removed from the cluster.
+     * @param retry            the number of times to retry a workflow if it fails
+     * @param timeout          total time to wait before the workflow times out
+     * @param pollPeriod       the poll interval to check whether a workflow completed or not
+     * @throws WorkflowResultUnknownException when the side affect of the operation
+     *                                        can't be determined
+     */
+    public void removeNode(@Nonnull String endpointToRemove, int retry,
```

Comment:
I am fine with this but I have to admit the lambda based logic here is way uglier than we had in the older style. It feels uncomfortable to debug.

Reply:
Yeah seems like runWorkflow should... take a workflow that defines these functions? This style makes sense when the lambdas are small...

Reply:
I reverted back to the previous verbose style. I'll attempt to address this verbosity in the next PR (i.e. force remove)